src/hotspot/share/gc/g1/g1ParScanThreadState.cpp

```diff
@@ -234,8 +243,9 @@ void G1ParScanThreadState::do_partial_array(PartialArrayScanTask task) {
     push_on_queue(ScannerTask(PartialArrayScanTask(from_obj)));
   }
 
-  HeapRegion* hr = _g1h->heap_region_containing(to_array);
-  G1ScanInYoungSetter x(&_scanner, hr->is_young());
+  G1HeapRegionAttr dest_attr = _g1h->region_attr(to_array);
+  assert(!dest_attr.is_in_cset(), "must not scan object from cset here");
+  G1SkipCardEnqueueSetter x(&_scanner, dest_attr.is_new_survivor() || dest_attr.is_in_cset());
```

Comment:
We just assert `!dest_attr.is_in_cset()`, why still `|| dest_attr.is_in_cset()`?

Reply:
`assert(!dest_attr.is_in_cset(), "must not scan object from cset here");` is because this should not happen (yet) as we do not split large objarrays that failed evacuation at this time. I'll remove that one, because something like "not implemented" or "should not happen" isn't great either.

`dest_attr.is_new_survivor() || dest_attr.is_in_cset()` is incorrect, just `dest_attr.is_new_survivor()` is correct (then again, it could not have happened yet).